tdms/tests/include/abstract_array_test_class.h

```diff
@@ -0,0 +1,204 @@
+   * @returns false matlab_input was not assigned, so no deletion was necessary
+   */
+  bool destroy_matlab_input() {
+    if (matlab_input_assigned) {
+      mxDestroyArray(matlab_input);
+      matlab_input_assigned = false;
+      return true;
+    } else {
+      return false;
+    }
+  }
+  /**
+   * @brief Creates a MATLAB structure array
+   *
+   * @param n_rows,n_cols Dimensions of the array. This is the number of identical structs in the array, not the size of a particular field
+   * @param n_fields Number of fields
+   * @param fields Names of the fields
+   */
+  void create_struct_array(int n_rows, int n_cols, int n_fields, const char *fields[]) {
+    matlab_input = mxCreateStructMatrix(n_rows, n_cols, n_fields, fields);
+    matlab_input_assigned = true;
+  }
+  /**
+   * @brief Creates a MATLAB structure array
+   *
+   * @param n_dimensions Number of dimensions of the array.
+   * @param dimensions Array or vector, containing the number of elements in each axis of the array
+   * @param n_fields Number of fields
+   * @param fields Names of the fields
+   */
+  void create_struct_array(int n_dimensions, int *dimensions, int n_fields, const char *fields[]) {
+    matlab_input = mxCreateStructArray(n_dimensions, (const mwSize *) dimensions, n_fields, fields);
+    matlab_input_assigned = true;
+  }
+  /**
+   * @brief Creates a 1-by-1 MATLAB structure array
+   *
+   * @param n_fields Number of fields
+   * @param fields Names of the fields
+   */
+  void create_1by1_struct(int n_fields, const char *fields[]) {
+    create_struct_array(1, 1, n_fields, fields);
+  }
+  /**
+   * @brief Creates the MATLAB empty struct (no fields, 0 size)
+   *
+   */
+  void create_empty_struct() {
+    create_struct_array(0, 1, 0, {});
+  }
+  /**
+   * @brief Creates a MATLAB numeric array
+   *
+   * @param n_dims Number of dimensions in the array
+   * @param dimensions Array/vector containing the number of elements in each dimension, sequentially
+   * @param data_type The MATLAB data type to populate the array with
+   * @param complex_flag Whether the data is real or complex
+   */
+  void create_numeric_array(int n_dims, int *dimensions, mxClassID data_type = mxDOUBLE_CLASS,
+                            mxComplexity complex_flag = mxREAL) {
+    matlab_input =
+            mxCreateNumericArray(n_dims, (const mwSize *) dimensions, data_type, complex_flag);
+    matlab_input_assigned = true;
+  }
+
+  /* Test functions to be overriden by each class.
+  These methods are designed to be overriden by the subclasses specific to each arrays.h class.
+
+  The default behaviour of each test_ method is to set ran_a_test to false - this means that subclasses that skip over certain functionality (EG do not need to test the result of providing an empty MATLAB input) are skipped in when run_all_class_tests() is run, and do not bloat the log.
+  If overriden, a test_ method should not alter ran_a_test
```

Comment:
Pretty much. Not all classes need to run every single test, so `AbstractArrayTest` has a private bool `ran_a_test` which is default-initialised to `true`. Then every `test_` function in `AbstractArrayTest` is a virtual method that by default sets `ran_a_test` to `false`.

The `AbstractArrayTest` class then has a method called `run_all_class_tests`, which DWYT. Subclasses of `AbstractArrayTest` can then override the `test_` functions relevant to their class definition, and leave the others alone. `run_all_class_tests` can then be inherited from `AbstractArrayTest` without the need to define it for _every_ subclass, since overridden `test_` methods will then not touch the `ran_a_test` bool, resulting in it remaining `true`, and indicating a test was run.

TLDR; it means that we only have to declare & define overrides for the `test_` methods that are relevant to each class, rather than doing this _plus_ redefining `run_all_class_tests` for every class (plus remembering to update this if we later decide to add more tests to a class).